sources/engine/Xenko.Assets/Entities/EntityHierarchyAssetBase.Upgraders.cs

```diff
@@ -56,5 +56,49 @@ protected override void UpgradeAsset(AssetMigrationContext context, PackageVersi
                 }
             }
         }
+
+        /// <summary>
+        /// Updates the Gravity field on all CharacterComponents in a SceneAsset from float to Vector3 to support three-dimensional gravity.
+        /// </summary>
+        protected class CharacterComponentGravityVector3Upgrader : AssetUpgraderBase
+        {
+            protected override void UpgradeAsset(AssetMigrationContext context, PackageVersion currentVersion, PackageVersion targetVersion, dynamic asset, PackageLoadingAssetFile assetFile, OverrideUpgraderHint overrideHint)
+            {
+                // Set up asset and entity hierarchy for reading.
+                var hierarchy = asset.Hierarchy;
+                var entities = (DynamicYamlArray)hierarchy.Parts;
+
+                // Loop through the YAML file.
+                foreach (dynamic entityDesign in entities)
+                {
+                    // Get the entity.
+                    var entity = entityDesign.Entity;
+
+                    // Save override and remove old Gravity element.
+                    var gravityOverride = entity.GetOverride("Gravity");
+                    entity.RemoveChild("Gravity");
+
+                    // Further loop to find CharacterComponents to upgrade.
+                    foreach (var component in entity.Components)
+                    {
+                        try
+                        {
+                            var componentTag = component.Value.Node.Tag;
+
+                            // Is this a character component?
+                            if (componentTag == "!CharacterComponent")
+                            {
+                                component.Value.SetOverride("Gravity", gravityOverride);
```

Comment:
Here is where the upgrade will happen:
* get the current float value for Gravity
* create a "Vector3" value (YAML version) and assign the previous value to its Y component.

Reply:
I assume I can get the "old value" from YAML by assigning `component.Value.PropertyName` (`PropertyName` is `Gravity` in this case) to a new `var`? This seems to be how it's done within `MoveRenderGroupToEntityUpgrader`, so I hope I'm correct in assuming this.

As for writing to YAML, I've not really worked with YAML serialisation before (although JSON is also ugly and I've worked with it), so would I initialise the `Vector3` representation as a string containing `oldGravity` as the Y component?

Reply:
Code should look like:
```
if (component.Value.ContainsChild("Gravity"))
{
    component.Value.Gravity = ...
}
```
For the actual float=>Vector3 transform code, take a look at `ProceduralModelAsset.Upgrader`, it contains an upgrader that does exactly that.

Reply:
@DW01 https://github.com/xenko3d/xenko/blob/releases/3.0.0.4/sources/engine/Xenko.Assets.Models/ProceduralModelAsset.cs#L70